### ISSUE TITLE
CAFE: Learning to Condense Dataset by Aligning Features

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#202]
# TL; DR

- Match the class-wise, layer-wise features of real data batch and synthetic (small) data batch while retaining class-level separability.

# Motivation

- Gradient-based dataset condensation methods are vulnerable to outliers that dominate the gradient.

# Formulation

<img width="1031" alt="image" src="https://user-images.githubusercontent.com/46308407/211787048-ab3b7c21-1fc0-44a9-a3d1-a5f2c69a2a9d.png">

### Feature Alignment

- Gather a batch of real data and synthetic data, all belonging to the same class.
- Forward them through the online network and obtain L-1 layer-wise features (excluding the final layer).
- Average them so that there are L-1 prototype-like features for each class.
- Simply compute MSE loss to match their distributions.

### Class Discrimination

- Grab the last layer features for both real and synthetic batch. (all classes)
- Get the logits by computing dot-product (similarity).
<img width="202" alt="image" src="https://user-images.githubusercontent.com/46308407/211788035-4877560b-5f3e-4c15-ba7c-26472e861e41.png">
- Assign cross-entropy loss to maintain discrimination capacity.
<img width="211" alt="image" src="https://user-images.githubusercontent.com/46308407/211788064-9f8b3157-c900-4af4-b574-c38c8b4b76d3.png">

### Bi-level optimization

- Outer Loop: update the synthetic datasets.
- Inner Loop: update model parameters.
- Rationale: synthetic sets are optimized for a wide spectrum of model parameters (note the features clearly depend on the model parameters).
- To make the optimization dynamic (hence robust to under-/over-fitting), keep a queue of classification accuracy and break each loop when it saturates.

<img width="421" alt="image" src="https://user-images.githubusercontent.com/46308407/211788780-01f940b9-2585-4ebb-b45e-aa7e301b9e5a.png">

# Evaluation

<img width="1031" alt="image" src="https://user-images.githubusercontent.com/46308407/211788644-6989df60-1daa-462f-9547-f1e6dbc902e9.png">
<img width="380" alt="image" src="https://user-images.githubusercontent.com/46308407/211788816-52924c2b-5411-4feb-bf5d-82366cae3733.png">
<img width="413" alt="image" src="https://user-images.githubusercontent.com/46308407/211788841-4ea3e6c3-50ed-421c-9b31-5195818c1016.png">
<img width="842" alt="image" src="https://user-images.githubusercontent.com/46308407/211788866-254dc48d-d189-44a5-8c21-e30a600a1756.png">
<img width="416" alt="image" src="https://user-images.githubusercontent.com/46308407/211788916-f656f5b4-1df7-43ea-b5fc-a8993a42f648.png">
<img width="856" alt="image" src="https://user-images.githubusercontent.com/46308407/211788996-1fb0f7d2-28d6-4ae9-a607-cf1dec9bc416.png">
<img width="422" alt="image" src="https://user-images.githubusercontent.com/46308407/211789040-ec304c61-8ccf-4e80-bd88-49027abad304.png">

# Inspiration

- Though feature matching is a rather classic idea, combining it with discrimination loss to strike the right balance was interesting.
- Keeping an accuracy queue to dynamically perform bi-level optimization was intuitive and proven effective.
- Evaluations are on relatively small scale benchmarks, and this approach still lacks scalability as we have to optimize synthetic set ***pixel-level***.
- Plus, this framework involves bi-level optimization, furthering incurring computational burden compared to #198 .
